ubuntu_image/classic_builder.py

```diff
@@ -0,0 +1,605 @@
+                # details.
+                Path(part_img).touch()
+                os.truncate(part_img, self.rootfs_size)
+            else:
+                run('dd if=/dev/zero of={} count=0 bs={} seek=1'.format(
+                    part_img, part.size))
+                if part.filesystem is FileSystemType.vfat:
+                    label_option = (
+                        '-n {}'.format(part.filesystem_label)
+                        # XXX I think this could be None or the empty string,
+                        # but this needs verification.
+                        if part.filesystem_label
+                        else '')
+                    # XXX: hard-coding of sector size.
+                    run('mkfs.vfat -s 1 -S 512 -F 32 {} {}'.format(
+                        label_option, part_img))
+            volume.part_images.append(part_img)
+            farthest_offset = max(farthest_offset, (part.offset + part.size))
+        # Calculate or check the final image size.
+        #
+        # XXX: Hard-codes last 34 512-byte sectors for backup GPT,
+        # empirically derived from sgdisk behavior.
+        calculated = ceil(farthest_offset / 1024 + 17) * 1024
+        if self.args.image_size is None:
+            volume.image_size = calculated
+        elif isinstance(self.args.image_size, int):
+            # One size to rule them all.
+            if self.args.image_size < calculated:
+                _logger.warning(
+                    'Ignoring image size smaller '
+                    'than minimum required size: vol[{}]:{} '
+                    '{} < {}'.format(
+                        i, name, self.args.given_image_size, calculated))
+                volume.image_size = calculated
+            else:
+                volume.image_size = self.args.image_size
+        else:
+            # The --image-size arguments are a dictionary, so look up the
+            # one used for this volume.
+            size_by_index = self.args.image_size.get(i)
+            size_by_name = self.args.image_size.get(name)
+            if size_by_index is not None and size_by_name is not None:
+                _logger.warning(
+                    'Ignoring ambiguous volume size; index+name given')
+                volume.image_size = calculated
+            else:
+                image_size = (size_by_index
+                              if size_by_name is None
+                              else size_by_name)
+                if image_size < calculated:
+                    _logger.warning(
+                        'Ignoring image size smaller '
+                        'than minimum required size: vol[{}]:{} '
+                        '{} < {}'.format(
+                            i, name, self.args.given_image_size, calculated))
+                    volume.image_size = calculated
+                else:
+                    volume.image_size = image_size
+
+    def prepare_filesystems(self):
```

Comment:
Just an idea for the future: since the prepare_filesystems, populate_filesystems and make_disk states are vaguely the same for ModelAssertionBuilder and ClassicBuilder, maybe we should consider creating a common class for these things (something like FilesystemAwareBuilder) that both would derive from? It's just a bit sad that we have to copy-paste these parts between both.

But as said, let's leave it as is for now and think about this in the future.

Reply:
Make sense.